test/pipelines/unit/test_connections.py

```diff
@@ -0,0 +1,611 @@
+            Dict[str, Mapping[str, Dict[str, Any]]],
+            Dict[str, Mapping[str, Dict[str, int]]],
+            id="nested-mapping-of-Any-values-to-nested-mapping-of-primitives",
+        ),
+        pytest.param(
+            Dict[str, Mapping[str, Dict[str, Any]]],
+            Dict[str, Mapping[str, Dict[str, TestClass1]]],
+            id="nested-mapping-of-Any-values-to-nested-mapping-of-classes",
+        ),
+        pytest.param(
+            Dict[str, Mapping[str, Dict[Any, Any]]],
+            Dict[str, Mapping[str, Dict[str, int]]],
+            id="nested-mapping-of-Any-keys-and-values-to-nested-mapping-of-primitives",
+        ),
+        pytest.param(
+            Dict[str, Mapping[str, Dict[Any, Any]]],
+            Dict[str, Mapping[str, Dict[str, TestClass1]]],
+            id="nested-mapping-of-Any-keys-and-values-to-nested-mapping-of-classes",
+        ),
+        pytest.param(
+            Literal["a", "b", "c"],
+            Literal["a", "b"],
+            id="different-literal-of-same-primitive",
+        ),
+        pytest.param(
+            Literal[TestEnum.TEST1],
+            Literal[TestEnum.TEST2],
+            id="different-literal-of-same-enum",
+        ),
+        pytest.param(
+            Tuple[Optional[Literal["a", "b", "c"]], Union[Path, Dict[int, TestClass1]]],
+            Tuple[Literal["a", "b", "c"], Union[Path, Dict[int, TestClass1]]],
+            id="deeply-nested-complex-type-is-compatible-but-cannot-be-checked",
+        ),
+    ],
+)
+def test_connect_non_compatible_types(output, input):
+    comp1 = make_component(output=output)
+    comp2 = make_component(input=input)
+
+    pipe = Pipeline()
+    pipe.add_component("c1", comp1)
+    pipe.add_component("c2", comp2)
+
+    with pytest.raises(PipelineConnectError, match="Cannot connect 'c1' with 'c2': no matching connections available."):
+        pipe.connect("c1", "c2")
+
+
+def test_connect_nonexisting_from_component():
+    add_1 = AddFixedValue()
+    add_2 = AddFixedValue()
+
+    pipe = Pipeline()
+    pipe.add_component("first", add_1)
+    pipe.add_component("second", add_2)
+    with pytest.raises(ValueError, match="Component named third not found in the pipeline"):
+        pipe.connect("third", "second")
+
+
+def test_connect_nonexisting_to_component():
+    add_1 = AddFixedValue()
+    add_2 = AddFixedValue()
+
+    pipe = Pipeline()
+    pipe.add_component("first", add_1)
+    pipe.add_component("second", add_2)
+    with pytest.raises(ValueError, match="Component named third not found in the pipeline"):
+        pipe.connect("first", "third")
+
+
+def test_connect_nonexisting_from_socket():
+    add_1 = AddFixedValue()
+    add_2 = AddFixedValue()
+
+    pipe = Pipeline()
+    pipe.add_component("first", add_1)
+    pipe.add_component("second", add_2)
+    with pytest.raises(PipelineConnectError, match="first.wrong does not exist"):
+        pipe.connect("first.wrong", "second")
+
+
+def test_connect_nonexisting_to_socket():
```

Comment:
I wouldn't use `input_socket` and `output_socket` because in a connect statement they are reversed, and it may get confusing (`connect_from` is an `OutputSocket` and `connect_to` is an `InputSocket`).

I will change these to `test_connect_to_does_not_existing` and follow the pattern